### ISSUE TITLE
Document SHOUTRRR_PARAMS: Your SHOUTRRR integration for generic webhooks only sends a string in the body when the default is application/json

### ISSUE THREAD

**Issue by HansBaumgaertner** [qdm12/ddns-updater#416]
Hi Quentin,

First off, long time user of your container with much happiness. 

Second, I have a bit of a problem that I managed to solve on my own, however for those that didn't go as far as I went to fix it, this bug report is to save them time.

I recently had a reason to use the notifications your container can generate with the SHOUTRRR integration. The idea was to grab the output from a generic webhook to trigger a Node-Red flow which would then update my custom-whitelist.yaml file with the updated dynamic IP address so I just would need restart my Crowdsec container for the changes to take affect. This proved to be a bit more difficult. 

Your documentation points everyone to go to the SHOUTRRR docs for the directions to properly setup the integrations for their supported services, however it is not mentioned that most of what you have exposed in your integration is centered around Gotify (which works for most users) and a half-integrated generic webhook. 

I and someone else spent many hours trying to figure out if it was Node-Red or Shoutrrr that was at fault for why we were only getting error 400 responses from the DDNS-Updater container when trying to send a generic webhook to a Node-Red endpoint. Node-Red's logs reported that there was a JSON error at position 0 with the character "L". 

To narrow down the problem more, I spun up a generic webhook container on my server to get the full message headers and body for troubleshooting. The results of that test concluded that the Content Type sent was application/json, but the body of the message was a string. This meant that Node-Red was expecting application/json, but was throwing an error due to receiving a string. No problem, the default output from SHOUTRRR is application/json, so all that needs to be done is to change the content type based on the documentation that you linked to. 

This is where the problem got worse. No matter if I included in the generic webhook URL extra parameters like ?contenttype=text/plain, it would ignore all of that. I tried searching endlessly on the SHOUTRRR forums thinking I had done a configuration problem. Nothing I threw at it seemed to work. 

So, since I ran out of ideas, I decided to open your code and take a look to see if there was something I could find that could help me narrow down what the problem was. I am not a Golang expert and to date have written 0 lines of go code. However, I was determined to find the problem. 

What I discovered was an undocumented env var in your code called SHOUTRRR_PARAMS which looked like it was the missing piece to my problem. Here is a screenshot of that bit taken from shoutrrr.go:
![image](https://user-images.githubusercontent.com/106229562/211776492-4232332f-ed80-46c8-a985-e8d9fe72662f.png)
Once I dropped in that env var into my compose file and set it to contenttype=text/plain, everything reported properly on the webhook tester app and then worked flawlessly in Node-Red. 

Moving forward, it would be good to have a look at your SHOUTRRR integration and documentation again. I only say this because of what I went through, but also since you have documented SHOUTRRR_ADDRESSES as a comma separated list of addresses to send to, but only with 1 configurable content type for the entire list. If I wanted to have included a webhook notification for Discord plus Node-Red, Discord would have failed since it expects a JSON formatted POST, but I had changed the content type to text/plain. Without changing the content type from the default, Discord would have not accepted a plain string in the body. 

Please have a look at my suggestions and if you need someone to test the updates, I would be happy to for you. Thanks!

**Comment by qdm12** [comment]
First of all, sorry for the many hours wasted, and this is definitely a nifty horrible bug and missing documentation.
Second, sorry for the massive delay, I was moving continents, and didn't have time to maintain much my projects.

See #411 as far as I have seen, sadly, I don't think you can configure a content type per service if you cannot do it using `SHOUTRRR_PARAMS`'; I even dug quite a bit in the Go shoutrrr library. So I'm thinking now to move everything to JSON, such that the default generic content type would work, what do you think?

**Comment by qdm12** [comment]
1. What should be the keys for the JSON message? Is this standardised, or is it specific to your service?
2. What plaintext do you get on your application? Just the message? Or also the title `DDNS updater`?

**Comment by qdm12** [comment]
I created https://github.com/containrrr/shoutrrr/issues/363 to get some clarifications. I still tend to think the default on shoutrrr should be `text/plain`, and, especially important if not, we need a way to specify the content type per service.

**Comment by piksel** [comment]
Sorry if I misunderstand something, but couldn't you just use:
```
SHOUTRRR_ADDRESSES="generic://<shoutrrr-webhook-endpoint-url>?template=json"
```
(for a json-encoded object, with `title` and `message`) or
```
SHOUTRRR_ADDRESSES="generic://<shoutrrr-webhook-endpoint-url>?contenttype=text/plain"
```
(for just the message as a string)

The `PARAMS` are used to set the title because we didn't add the `title` as a standardized field until later in shoutrrrrs lifetime, so it's a bit of a hack unfortunately. Everything you can put in `PARAMS`, you could just as well put in the URL to begin with.  
Everything in the params will be sent to all notification services, so it is probably not a good idea to expose it to the end users, since `title` is the only thing you can safely pass to all services without them complaining about unknown keys.

**Comment by qdm12** [comment]
> "how come you were not aware of such address options". sounds a bit offensive.

Not intended to be offensive at all 😉 I guess the `how come you were` could had been a `Why were you` maybe, but I still don't see it as offensive re-reading it 😸

> but does the SHOUTRRR_ADDRESSES take precedence over the environment parameter SHOUTRRR_PARAMS or would the title always be "DDNS Updater"

Interesting point, if it does take precedence (I hope it does), then I would rather remove SHOUTRRR_PARAMS and let the user add it to the address for each service, to avoid confusion which is, as we've seen, a lot more costly than copy pasting a parameter for each service. 

If it doesn't, perhaps a better idea would be to change SHOUTRRR_PARAMS to SHOUTRRR_TITLE to override the default title and avoid using global params 🤔

**Comment by piksel** [comment]
Params override the corresponding values from the address, yes.

>  I still think that it looks very clean using the additional environment variable,

For `generic`, yes, but for all of the "native" services it would probably not make much sense 🤷‍♀️.
The main problem with exposing is that it can be very confusing that all the fields needs to be available on all of the notification services.

> change SHOUTRRR_PARAMS to SHOUTRRR_TITLE to override the default title

That seems the most useful solution to me at least. Email (`smtp` service) is the only one that "needs" a title. E-mails without a subject are generally not great.

**Comment by qdm12** [comment]
What I coded locally:

1. `SHOUTRRR_PARAMS` is disabled and a warning redirects you to use `SHOUTRRR_DEFAULT_TITLE` and `SHOUTRRR_ADDRESSES` instead, if you still use it
2. The global default Shoutrrr title value defaults to `DDNS updater` if `SHOUTRRR_DEFAULT_TITLE` is left unset
3. After settings reading and validation, each shoutrrr address which does not have the `title` key in its url parameters has the keyvalue `title={default title}` added. I do this instead of using the global shoutrrr parameters to avoid overriding a title specified for a particular address. Note this also leaves the title untouched if an address has an empty title set like `?title=&otherkey=othervalue`. Finally, since this modification is done after settings parsing etc., it won't show up in each shoutrrr address logged in the settings tree at program start time.

Is everyone's happy with this change?

**Comment by qdm12** [comment]
Done in e4bb82d316a025619ced06c1f7a06802800f444f

TLDR: use `SHOUTRRR_TITLE` if you want to override the **default** title for every address (won't override the address specific title, if set).
